src/app/BufferedReadCallback.cpp

```diff
@@ -0,0 +1,235 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "lib/core/CHIPTLV.h"
+#include "lib/core/CHIPTLVTags.h"
+#include "lib/core/CHIPTLVTypes.h"
+#include "protocols/interaction_model/Constants.h"
+#include "system/SystemPacketBuffer.h"
+#include "system/TLVPacketBufferBackingStore.h"
+#include <app/BufferedReadCallback.h>
+#include <app/InteractionModelEngine.h>
+
+namespace chip {
+namespace app {
+
+void BufferedReadCallback::OnReportBegin(const ReadClient * apReadClient)
+{
+    mCallback.OnReportBegin(apReadClient);
+}
+
+void BufferedReadCallback::OnReportEnd(const ReadClient * apReadClient)
+{
+    CHIP_ERROR err = DispatchBufferedData(nullptr, mBufferedPath, StatusIB(), true);
+    if (err != CHIP_NO_ERROR)
+    {
+        mCallback.OnError(apReadClient, err);
+    }
+
+    mCallback.OnReportEnd(apReadClient);
+}
+
+CHIP_ERROR BufferedReadCallback::GenerateListTLV(System::PacketBufferTLVReader & aReader)
+{
+    System::PacketBufferHandle handle = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSizeWithoutReserve, 0);
+    System::PacketBufferTLVWriter writer;
+    TLV::TLVType outerType;
+
+    //
+    // Initialize the writer, but with chaining enabled, since we'll very likely
+    // exceed the size of a single packet buffer when reconstituing the entire list
+    //
+    writer.Init(std::move(handle), /* useChainedBuffers */ true);
+
+    ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Array, outerType));
+
+    for (auto & bufHandle : mBufferedList)
+    {
+        System::PacketBufferTLVReader reader;
+
+        reader.Init(std::move(bufHandle));
+
+        ReturnErrorOnFailure(reader.Next());
+        ReturnErrorOnFailure(writer.CopyElement(TLV::AnonymousTag, reader));
+    }
+
+    ReturnErrorOnFailure(writer.EndContainer(outerType));
+
+    writer.Finalize(&handle);
+    aReader.Init(std::move(handle));
```

Comment:
At this point `handle` is maybe-chained... but `PacketBufferTLVReader` does not handle chained buffers.  It will only look at the first buffer in the chain.

Perhaps we should add a ChainedPacketBufferTLVReader for use here?  Or rename PacketBufferTLVReader to SinglePacketBufferTLVReader, or something.